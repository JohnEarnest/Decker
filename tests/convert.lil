
# data conversion tests

t.a:11,22,33
t.b:0,1,0 t.t:("one,two","three\nfour","five")
t:table t
show[c:writecsv[t "fbss"]]
show[readcsv[c "fbss"]]
show[writecsv[table range 5]] # rectangular table(!)

t2:insert foo bar with
 "A" -2.953
 "B" 99.8
end
show[writecsv[t2 "sc"]]

show[readcsv["foo,bar\nA,-$2.953\nB,$99.8" "sc"]]     # currency casts
show[readcsv["A,B\none,-12.3\ntwo,45.999" "si"]]      # integer casts
show[readcsv["A,B\none,-12.3\ntwo,45.999"]]           # unspecified datatypes
show[readcsv["A,B\none\ntwo,45.999"]]                 # pad ragged column data
show[readcsv["A,B,C\none,11,true\ntwo,12,true" "si"]] # trim extra columns
show[readcsv["A,B\none,-12.3\ntwo,45.999\n"]]         # ignore a trailing newline
show[readcsv["A,B\none,-12.3\ntwo,45.999\n" "_i"]]    # explicitly drop column
show[readcsv["A,B\none,-12.3\ntwo,45.999" "sisi"]]    # pad out missing columns
show[readcsv[""]]                                     # no headers or data
show[readcsv["first,second,third"]]                   # headers only
show[readcsv["first,second,third\n  "]]               # headers only, with whitespace
show[readcsv["a,b\n,99\n,"]]                          # ragged last cell, no spec
show[readcsv["a,b\n,99\n," "si"]]                     # ragged last cell, spec

# custom delimiters
t3:insert foo bar with "C|D" 789 into t2
show[writecsv[t3 "sc" "|"]]
show[readcsv["first|second|third\n11|22|33" "fff" "|"]]

# special column types
show[t4:readcsv["number,fruit,bool\n1,Apple,1\n2,Banana,0" "ILB"]]
show[writecsv[t4 "ILB"]]
show[t4:readcsv["name,rich
one,\"<'text':['Red'],'font':['menu'],'arg':[''],'pat':[35]>\"
two,\"<'text':['Blue'],'font':['menu'],'arg':[''],'pat':[38]>\"
three,\"<'text':['Green'],'font':['menu'],'arg':[''],'pat':[41]>\"" "sT"]]
show[o4:writecsv[t4 "sT"]]
show[readcsv[o4 "sT"]~t4]

# allow spaces in column names
show[t5:readcsv["Multi Word,another\n-1337.5,\"quoted string\""]]
show[writecsv[t5]]

# handle 'j' columns
show[writecsv[insert name value with
 "Alpha" 11,22
 "Beta"  33,44
 "Gamma" 55,66
end "sj"]]
show[readcsv["name,value\nAlpha,\"[11,22]\"\nBeta,\"[33,44]\"\nGamma,\"[55,66]\"" "sj"]]
