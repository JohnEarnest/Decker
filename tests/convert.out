"a,b,t,c4\n11,false,\"one,two\",\n22,true,\"three\nfour\",\n33,false,five,"
+----+---+---------------+----+
| a  | b | t             | c4 |
+----+---+---------------+----+
| 11 | 0 | "one,two"     | "" |
| 22 | 1 | "three\nfour" | "" |
| 33 | 0 | "five"        | "" |
+----+---+---------------+----+
"value\n0\n1\n2\n3\n4"
"foo,bar\nA,-$2.95\nB,$99.80"
+-----+--------+
| foo | bar    |
+-----+--------+
| "A" | -2.953 |
| "B" | 99.8   |
+-----+--------+
+-------+-----+
| A     | B   |
+-------+-----+
| "one" | -12 |
| "two" | 45  |
+-------+-----+
+-------+----------+
| A     | B        |
+-------+----------+
| "one" | "-12.3"  |
| "two" | "45.999" |
+-------+----------+
+-------+----------+
| A     | B        |
+-------+----------+
| "one" | nil      |
| "two" | "45.999" |
+-------+----------+
+-------+----+
| A     | B  |
+-------+----+
| "one" | 11 |
| "two" | 12 |
+-------+----+
+-------+----------+
| A     | B        |
+-------+----------+
| "one" | "-12.3"  |
| "two" | "45.999" |
+-------+----------+
+-----+
| B   |
+-----+
| -12 |
| 45  |
+-----+
+-------+-----+-----+-----+
| A     | B   | c2  | c3  |
+-------+-----+-----+-----+
| "one" | -12 | nil | nil |
| "two" | 45  | nil | nil |
+-------+-----+-----+-----+
++
||
++
+-------+--------+-------+
| first | second | third |
+-------+--------+-------+
+-------+--------+-------+
| first | second | third |
+-------+--------+-------+
| ""    | nil    | nil   |
+-------+--------+-------+
+----+------+
| a  | b    |
+----+------+
| "" | "99" |
| "" | ""   |
+----+------+
+----+-----+
| a  | b   |
+----+-----+
| "" | 99  |
| "" | nil |
+----+-----+
"foo|bar\nA|-$2.95\nB|$99.80\n\"C|D\"|$789.00"
+-------+--------+-------+
| first | second | third |
+-------+--------+-------+
| 11    | 22     | 33    |
+-------+--------+-------+
+--------+----------+------+
| number | fruit    | bool |
+--------+----------+------+
| 1      | "Apple"  | 1    |
| 2      | "Banana" | 0    |
+--------+----------+------+
"number,fruit,bool\n1,Apple,true\n2,Banana,false"
+---------+------------------------------------------+
| name    | rich                                     |
+---------+------------------------------------------+
| "one"   | insert text font arg pat with "Red" "men |
| "two"   | insert text font arg pat with "Blue" "me |
| "three" | insert text font arg pat with "Green" "m |
+---------+------------------------------------------+
"name,rich\none,\"<\"\"text\"\":[\"\"Red\"\"],\"\"font\"\":[\"\"menu\"\"],\"\"arg\"\":[\"\"\"\"],\"\"pat\"\":[35]>\"\ntwo,\"<\"\"text\"\":[\"\"Blue\"\"],\"\"font\"\":[\"\"menu\"\"],\"\"arg\"\":[\"\"\"\"],\"\"pat\"\":[38]>\"\nthree,\"<\"\"text\"\":[\"\"Green\"\"],\"\"font\"\":[\"\"menu\"\"],\"\"arg\"\":[\"\"\"\"],\"\"pat\"\":[41]>\""
1
+------------+-----------------+
| Multi Word | another         |
+------------+-----------------+
| "-1337.5"  | "quoted string" |
+------------+-----------------+
"Multi Word,another\n-1337.5,quoted string"
"name,value\nAlpha,\"[11,22]\"\nBeta,\"[33,44]\"\nGamma,\"[55,66]\""
+---------+---------+
| name    | value   |
+---------+---------+
| "Alpha" | (11,22) |
| "Beta"  | (33,44) |
| "Gamma" | (55,66) |
+---------+---------+
